pllug-presentation-system/presentation-data/domdocumentdivider.cpp

```diff
@@ -0,0 +1,125 @@
+}
+
+QVector<QDomDocument> DomDocumentDivider::divideBySeparators(const QDomDocument &htmlDocument)
+{
+    QVector<QDomDocument> rParts;
+    QDomDocument currentPart;
+
+    QDomNode domNode = htmlDocument.documentElement().firstChild();
+    QDomElement domElement;
+    while(!(domNode.isNull()))
+    {
+        domElement = domNode.toElement();
+        if(!(domElement.isNull()))
+        {
+            qDebug() << "tag:" << domElement.tagName() << "text" << domElement.text();
+            if (isSeparator(domElement))
+            {
+                if(currentPart.childNodes().count() != 0)
+                {
+                    rParts.push_back(currentPart);
+                    currentPart = QDomDocument();
+                }
+            }
+            else
+            {
+                currentPart.appendChild(domNode);
+            }
+        }
+        domNode = domNode.nextSiblingElement();
+    }
+
+    if (currentPart.childNodes().count() != 0)
+    {
+        rParts.push_back(currentPart);
+    }
+
+    return rParts;
+}
+
+QVector<Slide> DomDocumentDivider::divideByHeaders(const QVector<QDomDocument> &parts)
+{
+    QVector<Slide> rSlides;
+    Slide currentSlide;
+
+    for (const QDomDocument &part: parts)
+    {
+        QDomNode domNode = part.documentElement().firstChild();
+        QDomElement domElement;
+        while(!(domNode.isNull()))
+        {
+            domElement = domNode.toElement();
+            if(!(domElement.isNull()))
+            {
+                qDebug() << "tag:" << domElement.tagName() << "text" << domElement.text();
+                if (isHeader(domElement))
+                {
+                    if (currentSlide.elementsCount() != 0)
+                    {
+                        rSlides.push_back(currentSlide);
+                        // FIXME: Needed constructor doesn't yet exist
+                        //                    currentSlide = Slide(domElement);
+                    }
+                }
+                else
+                {
+                    // FIXME: Method PresentationElementFactory::create() isn't yet implemented.
+                    //                    currentSlide.addElement(PresentationElementFactory().create(domElement.text(), domElement.tagName()));
+                }
+            }
+            domNode = domNode.nextSiblingElement();
+        }
+        if (currentSlide.elementsCount() != 0)
+        {
+            rSlides.push_back(currentSlide);
+        }
+    }
+
+    return rSlides;
+}
+
+Presentation *DomDocumentDivider::divideSlides(const QVector<Slide> &slides)
+{
+    Presentation *rPresentation = new Presentation();
+
+    for (const Slide &undividedSlide: slides)
+    {
+        QVector<Slide *> smallerSlides = divideSlide(undividedSlide, cMaxNumOfElements);
+        for (Slide *slide: smallerSlides)
+        {
+            rPresentation->appendSlide(std::make_unique<Slide>(*slide));
+        }
+    }
+
+    return rPresentation;
```

Comment:
I think, we can return std::unique_ptr here? as we only creates an object, but not takes an ownership.